### PR TITLE
migrate apache container to alpine

### PR DESCRIPTION
**Pull request by Zoey2936**
- [x] TODO: check if the www-data user has the same id
- [x] needs #1619

Signed-off-by: Zoey <zoey@z0ey.de>

### CODE REVIEW

**Comment by szaimen** [comment]
Wow, thanks a lot Zoey! I will include this in the release after 4.0.1 :)

File: Containers/apache/Dockerfile [hunk 1]
```diff
@@ -1,60 +1,40 @@
 # Caddy is a requirement
 FROM caddy:2.6.2-alpine as caddy
 
-FROM debian:bullseye-20221219-slim
+FROM httpd:2.4.54-alpine3.17
```
Comment (szaimen) [inline Containers/apache/Dockerfile:4]: the image needs to be change to the php-fpm-alpine

Comment (Zoey2936) [reply Containers/apache/Dockerfile:4]: why?

Comment (szaimen) [reply Containers/apache/Dockerfile:4]: ah sorry, wrong PR 😅

File: Containers/apache/Dockerfile [hunk 2]
```diff
@@ -1,59 +1,44 @@
...
```
Comment (szaimen) [inline Containers/apache/Dockerfile:4]: lets wait for alpine 3.17.1 as well

File: Containers/apache/Dockerfile [hunk 3]
```diff
@@ -1,59 +1,44 @@
...
+		-e 's/^#\(LoadModule .*mod_proxy_fcgi.so\)/\1/' \
+		-e 's/^#\(LoadModule .*mod_setenvif.so\)/\1/' \
+		-e 's/^#\(LoadModule .*mod_env.so\)/\1/' \
+		-e 's/^#\(LoadModule .*mod_mime.so\)/\1/' \
+		-e 's/^#\(LoadModule .*mod_dir.so\)/\1/' \
+		-e 's/^#\(LoadModule .*mod_authz_core.so\)/\1/' \
+		-e 's/^#\(LoadModule .*mod_alias.so\)/\1/' \
+		-e 's/^#\(LoadModule .*mod_mpm_event.so\)/\1/' \
+		-e 's/\(LoadModule .*mod_mpm_worker.so\)/#\1/' \
+		-e 's/\(LoadModule .*mod_mpm_prefork.so\)/#\1/' \
```
Comment (szaimen) [inline Containers/apache/Dockerfile:29]: I think prefork is not supported. See https://github.com/nextcloud/spreed/pull/8509/files

Comment (Zoey2936) [reply Containers/apache/Dockerfile:29]: Thats why we disable mpm_worker and mpm_prefork and enable mpm_event

Comment (szaimen) [reply Containers/apache/Dockerfile:29]: Ah I see. Sorry, didnt read correctly 😅

File: Containers/apache/Dockerfile (hunk 2 above)
Comment (Zoey2936) [reply Containers/apache/Dockerfile:4]: the httpd image was now pushed to alpine v3.17.1

Comment (szaimen) [reply Containers/apache/Dockerfile:4]: It looks like I will not get to this due to time problems before the next beta release which is planned for coming tuesday. Sorry for that! I'll try to include it in the version afterwards.

**Comment by szaimen** [comment]
I wonder: would it make sense to enable http2 in Apache? If I understand correctly is caddy enabling http2 and http3 out of the box for outgoing connections so our internal apache does not need it? Or WDYT?

**Comment by Zoey2936** [comment]
> I wonder: would it make sense to enable http2 in Apache? If I understand correctly is caddy enabling http2 and http3 out of the box for outgoing connections so our internal apache does not need it? Or WDYT?

It would make the connection between caddy and apache uses http2

We would need to configure it inside the caddy file: https://caddyserver.com/docs/caddyfile/directives/reverse_proxy#upstream-addresses (`reverse_proxy h2c://localhost:8000 {` instead of `reverse_proxy localhost:8000 {`)

But:
> versions allows customizing which versions of HTTP to support. As a special case, "h2c" is a valid value which will enable cleartext HTTP/2 connections to the upstream (however, this is a non-standard feature that does not use Go's default HTTP transport, so it is exclusive of other features; subject to change or removal). Default: 1.1 2, or if scheme is h2c://, h2c 2

we can only use h2c and not h2 since caddy connects to apache unencrypted if I understand correctly

[Also nginx hasn't implemented this because they don't see any sense in this.](https://mailman.nginx.org/pipermail/nginx/2015-December/049445.html)

**Comment by szaimen** [comment]
Ah I see but would this bring any performance advantage?

**Comment by Zoey2936** [comment]
> Ah I see but would this bring any performance advantage?

If I read the nginx mail, it could even degrade the performance

**Comment by szaimen** [comment]
> > Ah I see but would this bring any performance advantage?
> 
> If I read the nginx mail, it could even degrade the performance (maybe read it yourself)

I see. Okay good, then I'd say lets leave it as it is without http2 enabled in apache :)

**Comment by szaimen** [comment]
I'd like to keep the userid and maybe also the user and group-name. Can we add something like this here as well? https://github.com/nextcloud/all-in-one/blob/b249e02d53f063e1b75df96f50007a329d94717e/Containers/nextcloud/Dockerfile#L4-L11

File: Containers/apache/Dockerfile [hunk 4]
```diff
@@ -1,59 +1,44 @@
...
-        apache2 \
-        supervisor \
-        wget \
-        ca-certificates \
-        openssl \
-        netcat \
-        dpkg-dev \
-    ; \
-    rm -rf /var/lib/apt/lists/*
+RUN apk add --no-cache bash tzdata coreutils supervisor wget ca-certificates openssl netcat-openbsd
```
Comment (szaimen) [inline Containers/apache/Dockerfile:11]: can we keep the former formatting?

Comment (Zoey2936) [reply Containers/apache/Dockerfile]: done

**Comment by Zoey2936** [comment]
> I'd like to keep the userid and maybe also the user and group-name. Can we add something like this here as well?
> 
> https://github.com/nextcloud/all-in-one/blob/b249e02d53f063e1b75df96f50007a329d94717e/Containers/nextcloud/Dockerfile#L4-L11

we just need the www-data user/group with id 33?

File: Containers/apache/Dockerfile [hunk 5]
```diff
@@ -1,59 +1,60 @@
...
+		-e 's/^#\(LoadModule .*mod_mpm_event.so\)/\1/' \
+		-e 's/\(LoadModule .*mod_mpm_worker.so\)/#\1/' \
+		-e 's/\(LoadModule .*mod_mpm_prefork.so\)/#\1/' \
+		conf/httpd.conf
+
+COPY nextcloud.conf conf
+RUN echo Include conf/nextcloud.conf | tee -a conf/httpd.conf
+
+RUN sed -i "/^Listen /d" conf/httpd.conf
+echo "ServerName localhost" | tee -a conf/httpd.conf
```
Comment (szaimen) [inline Containers/apache/Dockerfile:52]: same as https://github.com/nextcloud/all-in-one/pull/1577/files#r1085253864 and below

Comment (Zoey2936) [reply Containers/apache/Dockerfile]: done

**Comment by Zoey2936** [comment]
@szaimen should be finished now

File: Containers/apache/start.sh [hunk 6]
```diff
@@ -48,7 +48,7 @@ echo "$CADDYFILE" > /Caddyfile
 # Add caddy path
 mkdir -p /mnt/data/caddy/
 
-# Fix apache sturtup
-rm -f /var/run/apache2/apache2.pid
+# Fix apache startup
+rm -f logs/httpd.pid
```
Comment (szaimen) [inline Containers/apache/start.sh:52]: Can we use an absloute path here?

Comment (Zoey2936) [reply Containers/apache/start.sh:52]: only there or also on all other locations?

Comment (szaimen) [reply Containers/apache/start.sh:52]: for now only here I'd say

Comment (Zoey2936) [reply Containers/apache/start.sh:52]:
```suggestion
rm -f /usr/local/apache2/logs/httpd.pid
```

File: Containers/apache/start.sh [hunk 7]
```diff
@@ -49,6 +49,6 @@ echo "$CADDYFILE" > /Caddyfile
 mkdir -p /mnt/data/caddy/
 
 # Fix apache startup
-rm -f logs/httpd.pid
+rm -f /usr/local/apache2/logs/httpd.pid
```
Comment (szaimen) [inline Containers/apache/start.sh:52]: btw, why is the pid located in logs?

Comment (Zoey2936) [reply Containers/apache/start.sh:52]: good question:

```console
root@mx:~# docker run --rm -it httpd:2.4.54-alpine3.17 httpd -V | grep pid
 -D DEFAULT_PIDLOG="logs/httpd.pid"
```

**Comment by szaimen** [comment]
LGTM now but will try to test this locally first 👍

**Review by szaimen — APPROVED** [review]
Builds correctly. Testing will show if everything works. Thank you very much! :)

**Comment by szaimen** [comment]
It works 🎉🎉🎉🎉

**Comment by Zoey2936** [comment]
> It works 🎉🎉🎉🎉

Perfect!

**Pull request merged**